### PR TITLE
check upload dir permissions

### PR DESCRIPTION
**Pull request by bbodenmiller**
Detect permission and ownership issues with the users uploads dir. This helps troubleshoot and correct [gitlab-org/omnibus-gitlab#311](https://gitlab.com/gitlab-org/omnibus-gitlab/issues/311). Fixes gitlabhq#7500, gitlabhq#7052. Related to gitlabhq#6281. Replaces #8769.

### CODE REVIEW

File: lib/tasks/gitlab/check.rake [hunk 1]
```diff
@@ -276,6 +277,45 @@ namespace :gitlab do
...
+          return
+        end
+
+        # if tmp upload dir has incorrect permissions, assume others do as well
+        if File.stat(upload_path_tmp).mode == 040755 && File.owned?(upload_path_tmp) # verify drwxr-xr-x permissions
+          puts "yes".green
+        else
+          puts "no".red
+          try_fixing_it(
+            "sudo chown -R #{gitlab_user} #{upload_path}",
```
Comment (stanhu) [inline lib/tasks/gitlab/check.rake]: Does the group name also need to be specified (e.g. `gitlab-www`)? See: https://gitlab.com/gitlab-org/omnibus-gitlab/issues/254

File: lib/tasks/gitlab/check.rake [hunk 2]
```diff
@@ -276,6 +277,45 @@ namespace :gitlab do
         fix_and_rerun
       end
     end
+    
+    def check_upload_perms
+      print "Upload directory permissions correct? ... "
+
+      upload_path = File.realpath(Rails.root.join('public/uploads'))
```
Comment (stanhu) [inline lib/tasks/gitlab/check.rake]: `realpath` will throw an exception if the directory does not exist.

File: lib/tasks/gitlab/check.rake (hunk 1 above)
Comment (bbodenmiller) [inline lib/tasks/gitlab/check.rake]: Non-Omnibus GitLab upload path does not have a group and Omnibus GitLab fixes the group with `gitlab-ctl reconfigure` per https://github.com/gitlabhq/gitlabhq/pull/8769#discussion_r24588605 so no need to check and fix it here. Also per https://gitlab.com/gitlab-org/omnibus-gitlab/issues/311#note_348792 you wouldn't want to run a recursive applying of the group anyways.

File: lib/tasks/gitlab/check.rake (hunk 2 above)
Comment (bbodenmiller) [inline lib/tasks/gitlab/check.rake]: Suggestions for a fix?

Comment (stanhu) [inline lib/tasks/gitlab/check.rake]: You can check that it exists and is a directory first via `File.directory?`.

File: lib/tasks/gitlab/check.rake (hunk 1 above)
Comment (stanhu) [inline lib/tasks/gitlab/check.rake]: Ok, sounds good.

File: lib/tasks/gitlab/check.rake (hunk 2 above)
Comment (bbodenmiller) [inline lib/tasks/gitlab/check.rake]: Fixed in latest commit.

**Comment by stanhu** [comment]
@jacobvosmaer, can you review this one?

File: lib/tasks/gitlab/check.rake (hunk 1 above)
Comment (jacobvosmaer) [inline lib/tasks/gitlab/check.rake]: Yeah everything under public/uploads belongs to 'git'. The git:gitlab-www ownership on public/uploads itself is just to give the gitlab-www group read(-only) access.

File: lib/tasks/gitlab/check.rake [hunk 3]
```diff
@@ -276,6 +277,57 @@ namespace :gitlab do
...
+          )
+          for_more_information(
+            see_installation_guide_section "GitLab"
+          )
+          fix_and_rerun
+        end
+      else
+        puts "no".red
+        try_fixing_it(
+          "sudo chmod 0750 #{upload_path}",
```
Comment (jacobvosmaer) [inline lib/tasks/gitlab/check.rake]: Where in the big if-else-if-else are we here? Is the file mode the only thing that's wrong in this code branch, or could the owner also be wrong?

**Comment by jacobvosmaer** [comment]
@bbodenmiller looks nice

@stanhu what is the thing about long argument lists you talk about [above](#commitcomment-12741005)? I don't see it in the code.

**Comment by stanhu** [comment]
Never mind; looks like the `-exec command {}` argument sends one command per filename; I thought `find` was bundling the command-line into one long string.

Looks good to me too.

File: lib/tasks/gitlab/check.rake (hunk 3 above)
Comment (bbodenmiller) [inline lib/tasks/gitlab/check.rake]: Are you proposing changing

`if File.stat(upload_path).mode == 040750`

to

`if File.stat(upload_path).mode == 040750 && File.owned?(upload_path)`?

**Comment by axilleas** [comment]
@jacobvosmaer @stanhu can we merge this then?

**Comment by jacobvosmaer** [comment]
Thanks @bbodenmiller !

**Pull request merged**